src/wallet/wallet.cpp

```diff
@@ -4093,3 +4095,82 @@ CKeyID CWallet::GetKeyForDestination(const CTxDestination& dest) const
     }
     return CKeyID();
 }
+
+static const std::string STYLE_STRING_NONE = "";
+static const std::string STYLE_STRING_LEGACY = "legacy";
+static const std::string STYLE_STRING_P2SH = "p2sh";
+static const std::string STYLE_STRING_SEGWIT = "segwit";
+
+OutputStyle ParseStyle(const std::string& style)
+{
+    if (style == STYLE_STRING_LEGACY) {
+        return STYLE_LEGACY;
+    } else if (style == STYLE_STRING_P2SH || style == "default") {
+        return STYLE_P2SH;
+    } else if (style == STYLE_STRING_SEGWIT) {
+        return STYLE_SEGWIT;
+    } else {
+        return STYLE_NONE;
+    }
+}
+
+
+const std::string& FormatStyle(OutputStyle style)
+{
+    switch (style) {
+    case STYLE_LEGACY: return STYLE_STRING_LEGACY;
+    case STYLE_P2SH: return STYLE_STRING_P2SH;
+    case STYLE_SEGWIT: return STYLE_STRING_SEGWIT;
+    default: assert(false);
+    }
+}
+
+std::vector<CTxDestination> CWallet::GetAllDestinationsForKey(const CPubKey& key)
+{
+    CKeyID keyid = key.GetID();
+    if (key.IsCompressed()) {
+        CTxDestination segwit = WitnessV0KeyHash(keyid);
+        CTxDestination p2sh = CScriptID(GetScriptForDestination(segwit));
+        return std::vector<CTxDestination>{std::move(keyid), std::move(p2sh), std::move(segwit)};
+    } else {
+        return std::vector<CTxDestination>{std::move(keyid)};
+    }
+}
+
+CTxDestination CWallet::GetDestinationForKey(const CPubKey& key, OutputStyle style)
+{
+    switch (style) {
+    case STYLE_LEGACY: return key.GetID();
+    case STYLE_P2SH:
+    case STYLE_SEGWIT: {
+        CTxDestination witdest = WitnessV0KeyHash(key.GetID());
+        CScript witprog = GetScriptForDestination(witdest);
+        // Check if the resulting program is solvable (i.e. doesn't use an uncompressed key)
+        if (!IsSolvable(witprog)) return key.GetID();
+        if (style == STYLE_P2SH) {
+            return CScriptID(witprog);
+        } else {
+            return witdest;
+        }
+    }
+    default: assert(false);
+    }
+}
+
+CTxDestination CWallet::GetDestinationForScript(const CScript& script, OutputStyle style)
+{
+    switch (style) {
+    case STYLE_LEGACY:
+    case STYLE_P2SH:
+        return CScriptID(script);
+    case STYLE_SEGWIT: {
+        CTxDestination witdest = WitnessV0ScriptHash(Hash(script.begin(), script.end()));
+        CScript witprog = GetScriptForDestination(witdest);
+        // Check if the resulting program is solvable (i.e. doesn't use an uncompressed key)
+        if (!IsSolvable(witprog)) return CScriptID(script);
+        AddCScript(witprog);
```

Comment:
Seems the caller should be responsible for adding this if it's a WitnessV0ScriptHash? Like the others, this doesn't seem like wallet functionality to me.

Reply:
Yes, it's ugly. I don't immediately see a nicer solution though, except for a callback to pass in, or duplicating part of this function just to find the correct program.

Note that if it weren't for the implicitly-know-about-witness-versions-of-keys logic, an equivalent line would be needed in GetDestinationForKey or its callers.